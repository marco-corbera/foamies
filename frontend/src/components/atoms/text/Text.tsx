import { twMerge } from "tailwind-merge";
import type { Size } from "@/types";

type Type = "h1" | "h2" | "h3" | "h4" | "p" | "span";
type Weight = "light" | "normal" | "semibold" | "bold" | "extrabold" | "black";

interface Props {
  weight?: Weight;
  uppercase?: boolean;
  italic?: boolean;
  size?: Size;
  type?: Type;
  children: string | React.ReactNode;
  className?: string;
}

const types: Record<Type, string> = {
  h1: "font-bold text-3xl md:text-4xl leading-10 md:leading-11",
  h2: "font-extrabold text-2xl md:text-3xl leading-7 md:leading-9",
  h3: "font-semibold text-xl md:text-2xl leading-5 md:leading-8",
  h4: "font-semibold text-xl md:text-2xl md:leading-8 leading-6 tracking-[0.35rem] md:tracking-[0.5rem]",
  p: "leading-normal text-md",
  span: "leading-none",
};

const fontWeight: Record<Weight, string> = {
  light: "font-light",
  normal: "font-normal",
  semibold: "font-semibold",
  bold: "font-bold",
  extrabold: "font-extrabold",
  black: "font-black",
};

const textSize: Record<Size, string> = {
  xs: "text-xs leading-1",
  sm: "text-sm leading-2",
  md: "text-md leading-3",
  lg: "text-lg leading-4",
  xl: "text-xl leading-4",
  "2xl": "text-2xl leading-7",
  "3xl": "text-3xl leading-10",
  "4xl": "text-4xl leading-11",
};

const Text = ({
  weight,
  uppercase = false,
  italic = false,
  size,
  children,
  type = "p",
  className = "",
}: Props) => {
  const Tag = type;

  const classes = twMerge(
    types[type],
    weight ? fontWeight[weight] : "",
    size && textSize[size],
    uppercase ? "uppercase" : "",
    italic ? "italic" : "",
    "col-span-full",
    "text-left",
    className,
  );
  return <Tag className={classes}>{children}</Tag>;
};

export const H1 = (props: Props) => <Text {...props} type="h1" />;
export const H2 = (props: Props) => <Text {...props} type="h2" />;
export const H3 = (props: Props) => <Text {...props} type="h3" />;
export const H4 = (props: Props) => <Text {...props} type="h4" />;
export const P = (props: Props) => <Text {...props} type="p" />;
export const Span = (props: Props) => <Text {...props} type="span" />;

export default Text;
