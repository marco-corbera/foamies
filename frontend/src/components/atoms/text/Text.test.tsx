import { render, screen } from "@testing-library/react";

import React from "react";

import { H1, H2, H3, H4, P, Span } from "./Text";

describe("Text Component Exports", () => {
  test("renders H1 with correct text content, tag, and snapshot", () => {
    render(<H1>This is an H1 test</H1>);
    const element = screen.getByText("This is an H1 test");
    expect(element).toBeInTheDocument();
    expect(element.tagName.toLowerCase()).toBe("h1");
    expect(element).toMatchSnapshot();
  });

  test("renders H2 with correct text content, tag, and snapshot", () => {
    render(<H2>This is an H2 test</H2>);
    const element = screen.getByText("This is an H2 test");
    expect(element).toBeInTheDocument();
    expect(element.tagName.toLowerCase()).toBe("h2");
    expect(element).toMatchSnapshot();
  });

  test("renders H3 with correct text content, tag, and snapshot", () => {
    render(<H3>This is an H3 test</H3>);
    const element = screen.getByText("This is an H3 test");
    expect(element).toBeInTheDocument();
    expect(element.tagName.toLowerCase()).toBe("h3");
    expect(element).toMatchSnapshot();
  });

  test("renders H4 with correct text content, tag, and snapshot", () => {
    render(<H4>This is an H4 test</H4>);
    const element = screen.getByText("This is an H4 test");
    expect(element).toBeInTheDocument();
    expect(element.tagName.toLowerCase()).toBe("h4");
    expect(element).toMatchSnapshot();
  });

  test("renders P with correct text content, tag, and snapshot", () => {
    render(<P>This is a P test</P>);
    const element = screen.getByText("This is a P test");
    expect(element).toBeInTheDocument();
    expect(element.tagName.toLowerCase()).toBe("p");
    expect(element).toMatchSnapshot();
  });

  test("renders Span with correct text content, tag, and snapshot", () => {
    render(<Span>This is a P test</Span>);
    const element = screen.getByText("This is a P test");
    expect(element).toBeInTheDocument();
    expect(element.tagName.toLowerCase()).toBe("span");
    expect(element).toMatchSnapshot();
  });
});
