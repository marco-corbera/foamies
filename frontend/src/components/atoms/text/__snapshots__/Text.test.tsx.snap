// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text Component Exports renders H1 with correct text content, tag, and snapshot 1`] = `
<h1
  class="font-bold text-3xl md:text-4xl leading-10 md:leading-11 col-span-full text-left"
>
  This is an H1 test
</h1>
`;

exports[`Text Component Exports renders H2 with correct text content, tag, and snapshot 1`] = `
<h2
  class="font-extrabold text-2xl md:text-3xl leading-7 md:leading-9 col-span-full text-left"
>
  This is an H2 test
</h2>
`;

exports[`Text Component Exports renders H3 with correct text content, tag, and snapshot 1`] = `
<h3
  class="font-semibold text-xl md:text-2xl leading-5 md:leading-8 col-span-full text-left"
>
  This is an H3 test
</h3>
`;

exports[`Text Component Exports renders H4 with correct text content, tag, and snapshot 1`] = `
<h4
  class="font-semibold text-xl md:text-2xl md:leading-8 leading-6 tracking-[0.35rem] md:tracking-[0.5rem] col-span-full text-left"
>
  This is an H4 test
</h4>
`;

exports[`Text Component Exports renders P with correct text content, tag, and snapshot 1`] = `
<p
  class="text-md col-span-full text-left"
>
  This is a P test
</p>
`;

exports[`Text Component Exports renders Span with correct text content, tag, and snapshot 1`] = `
<span
  class="leading-none col-span-full text-left"
>
  This is a P test
</span>
`;
